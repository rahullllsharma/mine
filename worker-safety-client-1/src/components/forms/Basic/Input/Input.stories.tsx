import type { ComponentMeta, ComponentStory } from "@storybook/react";
import type { Either } from "fp-ts/lib/Either";
import type { Errors } from "io-ts";
import * as tt from "io-ts-types";
import { initFormField } from "@/utils/formField";
import { Input } from "./Input";
import { TelephoneInput } from "./TelephoneInput";

export default {
  title: "JSB/Input",
  component: Input,
} as ComponentMeta<typeof Input>;

const decodeFieldInput: (input: string) => Either<Errors, tt.NonEmptyString> =
  tt.withMessage(tt.NonEmptyString, () => "Please provide a value").decode;

const field = initFormField(decodeFieldInput)("");

export const DefaultText: ComponentStory<typeof Input> = args => (
  <Input {...args} />
);

DefaultText.args = {
  type: "text",
  field,
  onChange: () => null,
  label: "First Name: *",
};

export const Telephone: ComponentStory<typeof TelephoneInput> = args => (
  <TelephoneInput {...args} />
);

Telephone.args = {
  field,
  onChange: () => null,
  label: "Telephone: *",
};
