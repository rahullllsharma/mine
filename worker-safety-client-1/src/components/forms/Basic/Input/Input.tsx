import type { FormField } from "@/utils/formField";
import type { IconName } from "@urbint/silica";
import type { Option } from "fp-ts/lib/Option";
import type * as t from "io-ts";
import * as O from "fp-ts/lib/Option";
import * as E from "fp-ts/lib/Either";
import cx from "classnames";
import { InputLayout } from "./InputLayout";

type CommonInputProps<V> = {
  id?: string;
  className?: string;
  label?: string;
  required?: boolean;
  disabled?: boolean;
  hasError?: boolean;
  htmlFor?: string;
  field: FormField<t.Errors, string, V>;
  onChange: (raw: string) => void;
  onKeyDown?: (e: React.KeyboardEvent<HTMLInputElement>) => void;
  placeholder?: string;
};

type InputDateTimeType = "date" | "time" | "datetime-local";
type InputDefaultType = "text" | "number";

export type InputProps<V> = CommonInputProps<V> &
  (
    | {
        type: InputDefaultType;
        icon?: IconName;
      }
    | {
        type: InputDateTimeType;
      }
  );

function getIcon<T>(props: InputProps<T>): Option<IconName> {
  switch (props.type) {
    case "text":
    case "number":
      return props.icon ? O.some(props.icon) : O.none;
    case "date":
    case "datetime-local":
      return O.some("calendar");
    case "time":
      return O.some("clock");
  }
}

function Input<V>(props: InputProps<V>) {
  const {
    id,
    className,
    type = "text",
    disabled,
    hasError = false,
    required = false,
    label,
    field,
    onChange,
    placeholder = "",
    onKeyDown,
  } = props;
  const { raw } = field;

  const icon = getIcon(props);

  return (
    <InputLayout
      id={id}
      className={className}
      label={label}
      field={field}
      disabled={disabled}
      icon={icon}
    >
      <input
        disabled={disabled}
        type={type}
        required={required}
        value={raw}
        aria-invalid={E.isLeft(field.val)}
        onChange={e => onChange(e.target.value)}
        onKeyDown={onKeyDown}
        className={cx(
          "flex-auto appearance-none rounded-md",
          "min-w-0 p-2 h-9",
          "focus:outline-none read-only:cursor-default",
          "disabled:bg-neutral-light-77 disabled:cursor-not-allowed disabled:opacity-38",
          { ["border-system-error-50"]: hasError }
        )}
        placeholder={placeholder}
      />
    </InputLayout>
  );
}

export { Input };
