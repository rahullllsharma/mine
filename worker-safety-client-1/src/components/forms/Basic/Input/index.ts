export * from "./Input";
export * from "./InputRaw";
export * from "./TelephoneInput";
