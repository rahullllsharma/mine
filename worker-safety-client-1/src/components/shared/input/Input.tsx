import type { PropsWithClassName } from "@/types/Generic";
import type { ForwardedRef, HTMLProps } from "react";
import type { IconName } from "@urbint/silica";
import { forwardRef } from "react";
import cx from "classnames";
import { Icon } from "@urbint/silica";
import ButtonIcon from "../button/icon/ButtonIcon";

// The Field* component will limit that, following the same principles defined on silica.
// Expects all normal props for defined for a HTMLInputElement
export type InputProps = HTMLProps<HTMLInputElement> &
  PropsWithClassName<{
    containerClassName?: string;
    error?: string;
    icon?: IconName;
    allowClear?: boolean;
    clearInputText?: () => void;
  }>;

const containerStaticStyles = `relative flex w-full items-center text-base font-normal text-neutral-shade-100 rounded-md border border-solid`;

function Input(
  {
    icon,
    className,
    containerClassName,
    type = "text",
    error = "",
    disabled,
    readOnly,
    placeholder,
    allowClear,
    clearInputText,
    ...props
  }: InputProps,
  ref: ForwardedRef<HTMLInputElement>
): JSX.Element {
  return (
    <div
      className={cx(containerStaticStyles, containerClassName, {
        ["border-system-error-40 focus-within:ring-system-error-40"]: !!error,
        ["border-neutral-shade-26 rounded-md focus-within:ring-1 focus-within:ring-brand-gray-60 bg-neutral-light-100"]:
          !error && !readOnly,
        ["bg-transparent border-none"]: readOnly,
      })}
    >
      {icon && !readOnly && (
        <Icon
          name={icon}
          className={cx(
            "ml-2 pointer-events-none w-6 h-6 text-xl leading-none bg-white",
            {
              ["opacity-38"]: disabled,
              ["text-neutral-shade-58"]: !disabled,
            }
          )}
        />
      )}
      <input
        {...props}
        ref={ref}
        aria-invalid={!!error}
        aria-errormessage={error}
        type={type}
        disabled={disabled}
        readOnly={readOnly}
        placeholder={readOnly ? undefined : placeholder}
        className={cx(
          "flex-auto rounded-md appearance-none focus:outline-none disabled:bg-neutral-light-77 read-only:cursor-default disabled:cursor-not-allowed disabled:opacity-38 min-w-0",
          className,
          {
            ["bg-transparent"]: readOnly,
            ["p-2"]: !readOnly,
          }
        )}
      ></input>
      {allowClear && props.value && (
        <ButtonIcon
          iconName="close_small"
          className={cx(
            "ml-2 pointer-events-auto w-6 h-6 text-xl leading-none bg-white",
            {
              ["opacity-38"]: disabled,
              ["text-neutral-shade-58"]: !disabled,
            }
          )}
          onClick={clearInputText}
        />
      )}
    </div>
  );
}

export default forwardRef(Input);
