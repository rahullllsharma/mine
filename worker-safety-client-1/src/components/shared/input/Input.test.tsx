import { render, screen } from "@testing-library/react";
import Input from "./Input";

describe(Input.name, () => {
  it("should render correctly", () => {
    const { asFragment } = render(<Input type="text" />);

    expect(asFragment()).toMatchSnapshot();
  });

  it("should render an icon when passed as prop", () => {
    const icon = "dashboard";
    const { asFragment } = render(<Input type="text" icon={icon} />);
    expect(
      document.querySelector("[aria-hidden]")?.classList.contains(`ci-${icon}`)
    ).toBeTruthy();

    expect(asFragment()).toMatchSnapshot();
  });

  it("should outline the input when has an error", () => {
    render(<Input type="text" error="some error" />);
    expect(screen.getByRole("textbox")).toHaveAttribute("aria-invalid", "true");
  });

  describe("when input is disabled", () => {
    it("should be disabled", () => {
      render(<Input type="text" disabled />);
      expect(screen.getByRole("textbox")).toBeDisabled();
    });

    it("should render the input text disabled", () => {
      const { asFragment } = render(<Input type="text" disabled />);

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe("when input is readOnly", () => {
    it("should be readonly and not have borders", () => {
      const { asFragment } = render(
        <Input type="text" readOnly defaultValue="Lorem ipsum" />
      );
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
