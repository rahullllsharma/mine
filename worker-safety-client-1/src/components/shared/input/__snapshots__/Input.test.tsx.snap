// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`undefined should render an icon when passed as prop 1`] = `
<DocumentFragment>
  <div
    class="relative flex w-full items-center text-base font-normal text-neutral-shade-100 rounded-md border border-solid border-neutral-shade-26 rounded-md focus-within:ring-1 focus-within:ring-brand-gray-60 bg-neutral-light-100"
  >
    <i
      aria-hidden="true"
      class="ci-dashboard ml-2 pointer-events-none w-6 h-6 text-xl leading-none bg-white text-neutral-shade-58"
    />
    <input
      aria-errormessage=""
      aria-invalid="false"
      class="flex-auto rounded-md appearance-none focus:outline-none disabled:bg-neutral-light-77 read-only:cursor-default disabled:cursor-not-allowed disabled:opacity-38 min-w-0 p-2"
      type="text"
    />
  </div>
</DocumentFragment>
`;

exports[`undefined should render correctly 1`] = `
<DocumentFragment>
  <div
    class="relative flex w-full items-center text-base font-normal text-neutral-shade-100 rounded-md border border-solid border-neutral-shade-26 rounded-md focus-within:ring-1 focus-within:ring-brand-gray-60 bg-neutral-light-100"
  >
    <input
      aria-errormessage=""
      aria-invalid="false"
      class="flex-auto rounded-md appearance-none focus:outline-none disabled:bg-neutral-light-77 read-only:cursor-default disabled:cursor-not-allowed disabled:opacity-38 min-w-0 p-2"
      type="text"
    />
  </div>
</DocumentFragment>
`;

exports[`undefined when input is disabled should render the input text disabled 1`] = `
<DocumentFragment>
  <div
    class="relative flex w-full items-center text-base font-normal text-neutral-shade-100 rounded-md border border-solid border-neutral-shade-26 rounded-md focus-within:ring-1 focus-within:ring-brand-gray-60 bg-neutral-light-100"
  >
    <input
      aria-errormessage=""
      aria-invalid="false"
      class="flex-auto rounded-md appearance-none focus:outline-none disabled:bg-neutral-light-77 read-only:cursor-default disabled:cursor-not-allowed disabled:opacity-38 min-w-0 p-2"
      disabled=""
      type="text"
    />
  </div>
</DocumentFragment>
`;

exports[`undefined when input is readOnly should be readonly and not have borders 1`] = `
<DocumentFragment>
  <div
    class="relative flex w-full items-center text-base font-normal text-neutral-shade-100 rounded-md border border-solid bg-transparent border-none"
  >
    <input
      aria-errormessage=""
      aria-invalid="false"
      class="flex-auto rounded-md appearance-none focus:outline-none disabled:bg-neutral-light-77 read-only:cursor-default disabled:cursor-not-allowed disabled:opacity-38 min-w-0 bg-transparent"
      readonly=""
      type="text"
      value="Lorem ipsum"
    />
  </div>
</DocumentFragment>
`;
